rust/actyx/ax-aql/src/language/type_check/drill_down_type.rs

```diff
@@ -0,0 +1,443 @@
+                },
+            ),
+            Ok(Type::Union(Arc::new((
+                Type::Atom(TypeAtom::Number(None)),
+                Type::NoValue
+            ))))
+        );
+
+        assert!(drill_down_type_by_index(
+            &Type::Array(Arc::new(Type::Atom(TypeAtom::Number(None)))),
+            &Ind {
+                head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                tail: vec![Index::String(String::from("some_string"))].try_into().unwrap(),
+            },
+        )
+        .is_err());
+    }
+
+    #[test]
+    fn dict() {
+        assert_eq!(
+            drill_down_type_by_index(
+                &Type::Dict(Arc::new(Type::Atom(TypeAtom::Number(None)))),
+                &Ind {
+                    head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                    tail: vec![Index::String(String::from("some_string"))].try_into().unwrap(),
+                },
+            ),
+            Ok(Type::Union(Arc::new((
+                Type::Atom(TypeAtom::Number(None)),
+                Type::NoValue
+            ))))
+        );
+        assert_eq!(
+            drill_down_type_by_index(
+                &Type::Dict(Arc::new(Type::Atom(TypeAtom::Number(None)))),
+                &Ind {
+                    head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                    tail: vec![Index::Number(1)].try_into().unwrap(),
+                },
+            ),
+            Ok(Type::Union(Arc::new((
+                Type::Atom(TypeAtom::Number(None)),
+                Type::NoValue
+            ))))
+        );
+    }
+
+    #[test]
+    fn record() {
+        assert_eq!(
+            drill_down_type_by_index(
+                &Type::Atom(TypeAtom::Record(
+                    vec![(
+                        Label::String("a".try_into().unwrap()),
+                        Type::Atom(TypeAtom::Record(
+                            vec![(
+                                Label::String("b".try_into().unwrap()),
+                                Type::Atom(TypeAtom::Record(
+                                    vec![(Label::Number(1), Type::Atom(TypeAtom::Null))].try_into().unwrap()
+                                ))
+                            )]
+                            .try_into()
+                            .unwrap()
+                        ))
+                    )]
+                    .try_into()
+                    .unwrap()
+                )),
+                &Ind {
+                    head: SimpleExpr::Variable(Var("_".to_string())).into(),
+                    tail: vec![
+                        Index::String("a".to_string()),
+                        Index::String("b".to_string()),
+                        Index::Number(1)
+                    ]
+                    .try_into()
+                    .unwrap(),
+                },
+            ),
+            Ok(Type::Atom(TypeAtom::Null))
+        );
+    }
+
+    #[test]
+    fn tuple() {
```

Comment:
Nit: I'd separate all the asserts in this test into multiple tests. AFAIK you also get to take more advantage of parallelism that way